### PR TITLE
[MI-2239][Server]: Written unit test cases for utils

### PR DESCRIPTION
**Pull request by avas27JTG**
Written unit test cases for utils

### CODE REVIEW

File: server/plugin/utils_test.go [hunk 1]
```diff
@@ -615,3 +617,101 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+			subscriptionsList: []*serializers.SubscriptionDetails{},
+			message:           fmt.Sprintf("No %s subscription exists", constants.CommandRepos),
+		},
+		{
+			description:       "ParseSubscriptionsToCommandResponse: empty boards subscription list",
+			channelID:         "mockChannelID",
+			createdBy:         "mockUser",
+			userID:            "mockUserID",
+			command:           constants.CommandBoards,
+			subscriptionsList: []*serializers.SubscriptionDetails{},
```
Comment (raghavaggarwal2308) [inline server/plugin/utils_test.go:708]: these fields are same in each testcase. we can use these directly

File: server/plugin/utils_test.go [hunk 2]
```diff
@@ -615,3 +617,101 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+			description:       "GetOffsetAndLimitFromQueryParams: invalid page query param",
+			project:           "mockProject",
+			queryParamPage:    "invalidNonIntegerString",
+			queryParamPerPage: "10",
+			expectedOffset:    0,
+			expectedLimit:     10,
+		},
+		{
+			description:       "GetOffsetAndLimitFromQueryParams: invalid per_page query param",
+			project:           "mockProject",
```
Comment (raghavaggarwal2308) [inline server/plugin/utils_test.go:659]: same

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/utils_test.go [hunk 3]
```diff
@@ -615,3 +617,101 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+	mockAPI := &plugintest.API{}
+	p.API = mockAPI
+	for _, testCase := range []struct {
+		description       string
+		subscriptionsList []*serializers.SubscriptionDetails
+		channelID         string
+		createdBy         string
+		userID            string
+		command           string
+		message           string
```
Comment (ayusht2810) [inline server/plugin/utils_test.go:691]:
```suggestion
		expectedMessage           string
```

File: server/plugin/utils_test.go [hunk 4]
```diff
@@ -615,3 +617,101 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+			description:       "ParseSubscriptionsToCommandResponse: empty repos subscription list",
+			channelID:         "mockChannelID",
+			createdBy:         "mockUser",
+			userID:            "mockUserID",
+			command:           constants.CommandRepos,
+			subscriptionsList: []*serializers.SubscriptionDetails{},
+			message:           fmt.Sprintf("No %s subscription exists", constants.CommandRepos),
+		},
+		{
+			description:       "ParseSubscriptionsToCommandResponse: empty boards subscription list",
```
Comment (ayusht2810) [inline server/plugin/utils_test.go:703]: don't we need to check cases when the repos or the boards list is not empty?

Comment (avas27JTG) [reply server/plugin/utils_test.go:703]: yes we can check but up to here we are able to meet(74%) our target of 70% coverage so didn't include that

**Review by raghavaggarwal2308 — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/utils_test.go [hunk 5]
```diff
@@ -615,3 +617,87 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+		{
+			description:       "GetOffsetAndLimitFromQueryParams: invalid per_page query param",
+			queryParamPage:    "1",
+			queryParamPerPage: "invalidNonIntegerString",
+			expectedOffset:    constants.DefaultPerPageLimit,
+			expectedLimit:     constants.DefaultPerPageLimit,
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"))
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:662]: Use something like this - https://github.com/Brightscout/mattermost-plugin-servicenow/blob/ae96b8fd1ae5a1da71575742894cd801fa55d3fa/server/testutils/mock.go#L7

File: server/plugin/utils_test.go [hunk 6]
```diff
@@ -615,3 +617,87 @@ func TestGetConnectAccountFirstMessage(t *testing.T) {
...
+
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:663]:
```suggestion
defer mockAPI.AssertExpectations(t)
```

File: server/plugin/utils_test.go (hunk 4 above)
Comment (manojmalik20) [reply server/plugin/utils_test.go:692]: If our coverage target is being met, we should not write unit tests for more functions, but if we are writing unit tests for a function, we should try to cover that function completely.

File: server/plugin/utils_test.go [hunk 7]
```diff
@@ -659,7 +662,10 @@ func TestGetOffsetAndLimitFromQueryParams(t *testing.T) {
 		},
 	} {
 		t.Run(testCase.description, func(t *testing.T) {
-			mockAPI.On("LogError", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"))
+			if testCase.queryParamPage != "1" && testCase.queryParamPerPage != "10" && testCase.expectedLimit != 10 && testCase.expectedOffset != 10 {
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:665]: There must be constants present for these values `10`.

File: server/plugin/utils_test.go [hunk 8]
```diff
@@ -694,9 +703,79 @@ func TestParseSubscriptionsToCommandResponse(t *testing.T) {
 			subscriptionsList: []*serializers.SubscriptionDetails{},
 			expectedMessage:   fmt.Sprintf("No %s subscription exists", constants.CommandBoards),
 		},
+		{
+			description:       "ParseSubscriptionsToCommandResponse: error in fetching filtered subscription list",
+			command:           constants.CommandBoards,
+			subscriptionsList: []*serializers.SubscriptionDetails{},
+			expectedMessage:   "Something went wrong, please try again later",
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:710]: There must be a constant already present for this error message.

File: server/plugin/utils_test.go [hunk 9]
```diff
@@ -694,9 +703,79 @@ func TestParseSubscriptionsToCommandResponse(t *testing.T) {
...
+					MattermostUserID: "mockUserID",
+					ServiceType:      constants.CommandBoards,
+					SubscriptionID:   "mockSubscriptionID",
+					OrganizationName: "mockOrganizationName",
+					ProjectName:      "mockProjectName",
+					EventType:        constants.SubscriptionEventWorkItemCreated,
+					CreatedBy:        "mockCreatedBy",
+					ChannelName:      "mockChannelName",
+				},
+			},
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:728]: Create a testutil function for returning this.

File: server/plugin/utils_test.go [hunk 10]
```diff
@@ -694,9 +703,79 @@ func TestParseSubscriptionsToCommandResponse(t *testing.T) {
...
+			createdBy:       constants.FilterCreatedByMe,
+			expectedMessage: fmt.Sprintf("###### %s subscription(s)\n| Subscription ID | Organization | Project | Event Type | Created By | Channel |\n| :-------------- | :----------- | :------ | :--------- | :--------- | :------ |\n| mockSubscriptionID | mockOrganizationName | mockProjectName | Work Item Created | mockCreatedBy | mockChannelName |\n", cases.Title(language.Und).String(constants.CommandBoards)),
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:730]: Same.

Comment (raghavaggarwal2308) [reply server/plugin/utils_test.go:730]: No constants present for this.....this string is printed in three different parts.....i combined them into one

File: server/plugin/utils_test.go (hunk 7 above)
Comment (raghavaggarwal2308) [reply server/plugin/utils_test.go:665]: These are just some mock values

File: server/plugin/utils_test.go [hunk 11]
```diff
@@ -707,65 +707,30 @@ func TestParseSubscriptionsToCommandResponse(t *testing.T) {
...
-					OrganizationName: "mockOrganizationName",
-					ProjectName:      "mockProjectName",
-					EventType:        constants.SubscriptionEventWorkItemCreated,
-					CreatedBy:        "mockCreatedBy",
-					ChannelName:      "mockChannelName",
-				},
-			},
-			createdBy:       constants.FilterCreatedByMe,
-			expectedMessage: fmt.Sprintf("No %s subscription exists", constants.CommandBoards),
+			description:       "ParseSubscriptionsToCommandResponse: no subscriptions created by the user is present",
```
Comment (manojmalik20) [inline server/plugin/utils_test.go:729]:
```suggestion
			description:       "ParseSubscriptionsToCommandResponse: no subscriptions created by the user are present",
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**